cpp/leaky_core.h

```diff
@@ -51,25 +48,21 @@ class TermState {
   }
 };
 
-void SignalState_AfterForkingChild();
-
-class SignalState {
+// An abstract type that approximates python's Callable-based signal().
+class SignalHandler : public Obj {
  public:
-  SignalState() {
-  }
-  void InitShell() {
-  }
-  void AddUserTrap(int sig_num, builtin_trap::_TrapHandler* handler) {
-    NotImplemented();
+  // XXX: This will probably break the garbage collector. We may need to rework
+  // this when it gets enabled.
+  SignalHandler() : Obj(Tag::FixedSize, kZeroMask, 0) {
   }
-  void RemoveUserTrap(int sig_num) {
-    NotImplemented();
-  }
-  int last_sig_num = 0;
-
-  DISALLOW_COPY_AND_ASSIGN(SignalState)
+  virtual void Run(int) = 0;
 };
 
+void Sigaction(int sig_num, SignalHandler* handler);
```

Comment:
I _think_ I like this pattern. It will require **a lot** of discipline to make sure we keep the python handlers signal-safe, though...

Reply:
see comment above -- the Python handlers definitely won't be signal safe, as running them will allocate memory, collect garbage, etc.